src/state/container/mod.rs

```diff
@@ -58,43 +58,82 @@ impl State {
     }
 
     /// Inserts new state, overriding existing state.
-    pub fn insert<T: CustomState>(&mut self, state: T) {
+    pub fn insert<T: CustomState<'a>>(&mut self, state: T) {
         self.map.insert(state);
     }
 
+    /// Tires to find an inner map containing T.
+    fn find<T: CustomState<'a>>(&self) -> Option<&Self> {
+        if self.map.has::<T>() {
+            Some(self)
+        } else {
+            self.parent()
+        }
+    }
+
+    /// Tires to find an inner map containing T.
+    fn find_mut<T: CustomState<'a>>(&mut self) -> Option<&mut Self> {
+        if self.map.has::<T>() {
+            Some(self)
+        } else {
+            self.parent_mut()
+        }
+    }
+
     /// Checks whether the state exists.
-    pub fn has<T: CustomState>(&self) -> bool {
-        self.map.has::<T>() || self.parent().map(|p| p.has::<T>()).unwrap_or_default()
+    pub fn has<T: CustomState<'a>>(&self) -> bool {
+        self.find::<T>().is_some()
     }
 
     /// Panics if the state does not exist.
     ///
     /// This is the recommended way to ensure the state
     /// is available in [Component::initialize](crate::framework::components::Component::initialize).
     #[track_caller]
-    pub fn require<T: CustomState>(&self) {
+    pub fn require<T: CustomState<'a>>(&self) {
         assert!(
             self.has::<T>(),
             "operator requires {} state",
             std::any::type_name::<T>()
         );
     }
 
+    /// Removes `T` from state and returns it.
+    ///
+    /// If `T` should only be removed temporarily, consider using [State::holding] instead.
+    #[track_caller]
+    pub fn take<T: CustomState<'a>>(&mut self) -> T {
+        self.find_mut::<T>().unwrap().map.take::<T>()
+    }
+
+    /// Access `T` mutably without borrowing the state.
+    ///
+    /// To make this possible, `T` will be removed temporarily.
+    /// This should only be used, if the state has to be passed to another function,
+    /// whilst borrowing from it.
+    #[track_caller]
+    pub fn holding<T: CustomState<'a>>(&mut self, code: impl FnOnce(&mut T, &mut Self)) {
+        let state_with_t = self.find_mut::<T>().unwrap() as *mut Self;
+
+        let mut instance = self.take::<T>();
+        code(&mut instance, self);
+
+        // Insert the state where it was before.
+        // This is save, because inner states can not be removed.
+        unsafe { state_with_t.as_mut().unwrap().insert(instance) };
```

Comment:
What exactly was the issue here that we need `unsafe`?

Reply:
I've changed it to a solution without `unsafe`.